### ISSUE TITLE
Stack overflow in multi-threaded mode during type instantiation

### ISSUE THREAD

**Issue by shahmir-oscilar** [microsoft/typescript-go#3081]
## Bug Report

`tsgo` crashes with a goroutine stack overflow when running in default (multi-threaded) mode on a large React/TypeScript codebase. The same codebase type-checks successfully with `--singleThreaded` and with standard `tsc`.

### tsgo Version

`7.0.0-dev.20260312.1` (also reproduces on `20260310.1` and `20260311.1`)

### Environment

- macOS, Apple Silicon (Darwin arm64)
- tsconfig: `strict: true`, `skipLibCheck: true`, `moduleResolution: "bundler"`, `noEmit: true`

### Reproduction

```
npx @typescript/native-preview@7.0.0-dev.20260312.1   # crashes
npx @typescript/native-preview@7.0.0-dev.20260312.1 --singleThreaded  # works
```

The project is a large React 18 + TypeScript codebase (~1600+ source files) using MUI, React Hook Form, Yup, React Query, and other libraries with deeply recursive/conditional generic types.

### Bisection

- **`20260309.1` and earlier**: No crash in multi-threaded mode (but `skipLibCheck` was not working correctly — errors were reported from `.d.ts` files in `node_modules`)
- **`20260310.1` and later**: Crashes in multi-threaded mode. `skipLibCheck` works correctly, and `--singleThreaded` works fine.

It appears the fix for `skipLibCheck` on March 10 may have introduced a concurrent type-checking code path that leads to unbounded recursion in multi-threaded mode.

### Stack Trace

```
runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow

goroutine 65279 [running]:
github.com/microsoft/typescript-go/internal/checker.(*Checker).instantiateTypeWithAlias(...)
  checker.go:21460
github.com/microsoft/typescript-go/internal/checker.(*Checker).instantiateType(...)
  checker.go:21457
github.com/microsoft/typescript-go/internal/checker.instantiateList[...](...)
  checker.go:22139
github.com/microsoft/typescript-go/internal/checker.(*Checker).instantiateTypes(...)
  checker.go:22122
github.com/microsoft/typescript-go/internal/checker.(*Checker).instantiateTypeWorker(...)
  checker.go:21604
github.com/microsoft/typescript-go/internal/checker.(*Checker).instantiateTypeWithAlias(...)
  checker.go:21489
github.com/microsoft/typescript-go/internal/checker.(*CompositeTypeMapper).Map(...)
  mapper.go:259
github.com/microsoft/typescript-go/internal/checker.(*Checker).getConditionalTypeInstantiation(...)
  checker.go:21845
github.com/microsoft/typescript-go/internal/checker.(*Checker).instantiateTypeWorker(...)
  checker.go:21628
github.com/microsoft/typescript-go/internal/checker.(*Checker).instantiateTypeWithAlias(...)
  checker.go:21489
github.com/microsoft/typescript-go/internal/checker.(*Checker).getTypeOfMappedSymbol(...)
  checker.go:20353
github.com/microsoft/typescript-go/internal/checker.(*Checker).getTypeOfSymbol(...)
  checker.go:15901
github.com/microsoft/typescript-go/internal/checker.(*Checker).getNonMissingTypeOfSymbol(...)
  checker.go:15925
github.com/microsoft/typescript-go/internal/checker.(*NodeBuilderImpl).addPropertyToElementList(...)
  nodebuilderimpl.go:2196
github.com/microsoft/typescript-go/internal/checker.(*NodeBuilderImpl).createTypeNodesFromResolvedType(...)
  nodebuilderimpl.go:2375
```

The recursion cycle is: `instantiateTypeWithAlias` → `instantiateTypeWorker` → `getConditionalTypeInstantiation` → `CompositeTypeMapper.Map` → `instantiateType` → `instantiateTypeWithAlias` (repeat).

### Expected Behavior

`tsgo` should complete type-checking without crashing, same as `--singleThreaded` mode.

### Workaround

Using `--singleThreaded` avoids the crash.

**Comment by jakebailey** [comment]
> It appears the fix for skipLibCheck on March 10 may have introduced a concurrent type-checking code path that leads to unbounded recursion in multi-threaded mode.

Which fix are you referring to?

**Comment by shahmir-oscilar** [comment]
> > It appears the fix for skipLibCheck on March 10 may have introduced a concurrent type-checking code path that leads to unbounded recursion in multi-threaded mode.
> 
> Which fix are you referring to?

Sorry, that was speculative. On my codebase, if I run:

`pnpx @typescript/native-preview@7.0.0-dev.20260309.1`, (same with --singleThreaded) I see errors from node_modules:

```
node_modules/.pnpm/gl-matrix@3.4.3/node_modules/gl-matrix/index.d.ts:93:15 - error TS1540: A 'namespace' declaration should not be declared using the 'module' keyword. Please use the 'namespace' keyword instead.

...
2897 export module vec2 {
                   ~~~~

node_modules/.pnpm/immutable@3.8.2/node_modules/immutable/dist/immutable-nonambient.d.ts:124:17 - error TS1540: A 'namespace' declaration should not be declared using the 'module' keyword. Please use the 'namespace' keyword instead.
...

Found 29 errors in 2 files.
```

but when I run `pnpx @typescript/native-preview@7.0.0-dev.20260310.1 --singleThreaded`, I don't see those errors

On that same version, removing `--singleThreaded` causes the stack overflow, and I'm able to reproduce the same thing on the latest version also

**Comment by jakebailey** [comment]
Yes, we got rid of that error. I don't think that's related.

**Comment by ahejlsberg** [comment]
Hmm, the `checker.(*NodeBuilderImpl)` on the stack means we're serializing something, probably because we're building a diagnostic message. Can you post the bottom 100 and top 100 frames of the stack?

**Comment by shahmir-oscilar** [comment]
@ahejlsberg here's the entire stack

[tsgo_output.txt](https://github.com/user-attachments/files/25961286/tsgo_output.txt)

**Comment by ahejlsberg** [comment]
It appears that `getTypeOfMappedSymbol` is attempting to report a circularity error, calling `TypeToString` to show the mapped type. That invokes the node builder's `TypeToTypeNode` which tries to serialize the properties of the mapped type and then sets off an infinite loop by attempting to get the symbol of the same mapped type.

I'll think about the most effective way to limit this type of circularity. Meanwhile, it would be awesome to have a repro if there's any way you can provide that.

**Comment by shahmir-oscilar** [comment]
Sorry, wasn't able to find root cause / come up with a sharable repro, but totally willing to try out any fixes y'all do to verify that it's fixed.

**Comment by jakebailey** [comment]
If you're willing to build from source, can you try #3093?

**Comment by shahmir-oscilar** [comment]
> If you're willing to build from source, can you try [#3093](https://github.com/microsoft/typescript-go/pull/3093)?

Fixed the stack overflow issue, but there's a discrepancy in the results of `--singleThreaded` vs multi-threaded

`--singleThreaded` still doesn't produce any errors, while multi-threaded now tells me:

```
❯ ../typescript-go/built/local/tsgo
src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''array_schema'' circularly references itself in mapped type '{ [K in keyof ArraySchemaAO]-?: PathImpl<K, ArraySchemaAO[K], ArraySchemaAO | IntegrationModuleAO | SchemaFieldAO | IntegrationFormData[K]>; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''array_schema'' circularly references itself in mapped type '{ [K in keyof ArraySchemaAO]-?: PathImpl<K, ArraySchemaAO[K], ArraySchemaAO | SchemaFieldAO | IntegrationFormData[K]>; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''array_schema'' circularly references itself in mapped type '{ [K in keyof ArraySchemaAO]-?: PathImpl<K, ArraySchemaAO[K], IntegrationFeatureAO[] | ArraySchemaAO | IntegrationFeatureAO | SchemaFieldAO | IntegrationFormData[K]>; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''array_schema'' circularly references itself in mapped type '{ [K in keyof ArraySchemaAO]-?: PathImpl<K, ArraySchemaAO[K], SchemaFieldAO[] | ArraySchemaAO | SchemaAO | SchemaFieldAO | IntegrationFormData[K]>; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''array_schema'' circularly references itself in mapped type '{ name: "name"; alias: "alias"; field_type: "field_type"; array_schema: any; object_schema: "object_schema" | (true extends true | AnyIsEqual<IntegrationFormData[K], SchemaAO> ? "object_schema" : "object_schema" | `object_schema.${...}`); ... 10 more ...; readable_roles: "readable_roles" | ...; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''array_schema'' circularly references itself in mapped type '{ name: "name"; alias: "alias"; field_type: "field_type"; array_schema: any; object_schema: any; default_value: "default_value"; optional: "optional"; description: "description"; enums: "enums" | ...; ... 6 more ...; readable_roles: "readable_roles" | ...; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''array_schema'' circularly references itself in mapped type '{ type: "type"; array_schema: any; object_schema: any; enums: "enums" | (true extends AnyIsEqual<IntegrationFormData[K], string[]> ? "enums" : "enums" | `enums.${number}`); tags: "tags" | ...; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''object_schema'' circularly references itself in mapped type '{ [K in keyof ArraySchemaAO]-?: PathImpl<K, ArraySchemaAO[K], SchemaFieldAO[] | ArraySchemaAO | SchemaAO | SchemaFieldAO | IntegrationFormData[K]>; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''object_schema'' circularly references itself in mapped type '{ [K in keyof SchemaFieldAO]-?: PathImpl<K, SchemaFieldAO[K], SchemaFieldAO[] | SchemaAO | SchemaFieldAO | IntegrationFormData[K]>; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''object_schema'' circularly references itself in mapped type '{ name: "name"; alias: "alias"; field_type: "field_type"; array_schema: "array_schema" | (true extends AnyIsEqual<IntegrationFormData[K], ArraySchemaAO> ? "array_schema" : "array_schema" | `array_schema.${any}`); ... 11 more ...; readable_roles: "readable_roles" | ...; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''object_schema'' circularly references itself in mapped type '{ name: "name"; alias: "alias"; field_type: "field_type"; array_schema: any; object_schema: any; default_value: "default_value"; optional: "optional"; description: "description"; enums: "enums" | ...; ... 6 more ...; readable_roles: "readable_roles" | ...; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~

src/components/integration/rest-api-integration-form.tsx:176:7 - error TS2615: Type of property ''object_schema'' circularly references itself in mapped type '{ type: "type"; array_schema: any; object_schema: any; enums: "enums" | (true extends AnyIsEqual<IntegrationFormData[K], string[]> ? "enums" : "enums" | `enums.${number}`); tags: "tags" | ...; }'.

176       formMethods.setValue('configuration.additional_headers', newHeaders, {
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
177         shouldDirty: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
178       });
    ~~~~~~~~


Found 12 errors in the same file, starting at: src/components/integration/rest-api-integration-form.tsx:176
``` 

`const formMethods = useFormContext<IntegrationFormData>();`, and `useFormContext` comes from `react-hook-form`

**Comment by shahmir-oscilar** [comment]
Some typing info, that hopefully helps:

```
type OmitStrict<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
```

```
export type IntegrationFormData = OmitStrict<IntegrationAO, 'dw_tables'> & {
  dw_tables?: Array<DWTableWithSchedule>;
  permission_tables: PermissionTable[];
};
```

```
export interface SchemaFieldAO {
...
    /**
     * 
     * @type {ArraySchemaAO}
     * @memberof SchemaFieldAO
     */
    'array_schema'?: ArraySchemaAO;
    /**
     * 
     * @type {SchemaAO}
     * @memberof SchemaFieldAO
     */
    'object_schema'?: SchemaAO;
...
}
```

```
export interface SchemaAO {
    /**
     * 
     * @type {Array<SchemaFieldAO>}
     * @memberof SchemaAO
     */
    'fields': Array<SchemaFieldAO>;
}
```

**Comment by ahejlsberg** [comment]
Happy to hear the stack overflow is fixed. With respect to the errors that only occur with concurrent checking, it's really hard to say what's going on without a repro. In single-threaded mode, all files are processed by one checker, whereas in concurrent mode, multiple checkers each process a subset of the files. This may cause interdependent type relations to be processed in different orders, which in turn could cause errors to not surface because recursion limiters stop some type exploration. We've worked hard to eliminate such differences (e.g. with the new total type ordering), but we also know that because we permit circularities and have recursion limiters, we can never totally eliminate differences.

That said, if you can provide a repro, I'd be happy to explore further.

**Comment by shahmir-oscilar** [comment]
During the process of trying to create an independent repro, I ended up finding that it's most likely the `--singleThreaded` _not_ returning the 12 errors is the issue. As I started chipping away and reducing the code in a couple of files, the same `circularly references itself in mapped type` error started showing up for `--singleThreaded` also

Going to try to get the repro done so that `--singleThreaded` (incorrectly, I believe) doesn't show the issue, and multi-threaded does.

Update 1: The circular error _doesn't_ show up in TS 5.9 or TS 6.0 beta in our actual code base, but does in the repro I was trying to create, which indicates that TS 7 is actually more correct

Update 2: According to https://github.com/react-hook-form/react-hook-form/issues/4055, this is apparently an issue with older versions of react-hook-form and on our end, we just need to upgrade to the latest version.

Honestly, with the stack overflow fixed, probably not worth investigating more on your end. Feel free to mark this as closed.

Thanks!

**Comment by jakebailey** [comment]
> The circular error doesn't show up in TS 5.9 or TS 6.0 beta in our actual code base

Try running 6.0 with `--stableTypeOrdering`. That might reveal the error.

**Comment by shahmir-oscilar** [comment]
> > The circular error doesn't show up in TS 5.9 or TS 6.0 beta in our actual code base
> 
> Try running 6.0 with `--stableTypeOrdering`. That might reveal the error.

`pnpm --package=typescript@6.0.1-rc dlx tsc --stableTypeOrdering` showed no errors.

<img width="637" height="132" alt="Image" src="https://github.com/user-attachments/assets/d38beae7-b5d8-4917-8f3d-ca270622857d" />
